### PR TITLE
support interface parameter for loading XDP program in netsh ebpf con…

### PR DESCRIPTION
**Pull request by shankarseal**
This is first part of changes for #754 .

The PR:
- adds support for `interface` parameter in `netsh ebpf add program` and `netsh ebpf set program` commands.
- adds netsh unit tests for the new parameter.
- FIxes a bug in `ebpf_api.cpp` module.

### CODE REVIEW

File: tools/netsh/ebpfnetsh.rc [hunk 1]
```diff
@@ -69,6 +69,11 @@ BEGIN
 \n                  determine the program type and expected attach\
 \n                  type.  If not specified, the default is xdp.\
 \n      pinpath   - Optionally the name to pin the program to.\
+\n      interface - Optionally the friendly name or index of a \
+\n                  network interface to which an XDP program should be attached. \
+\n                  If the interface is not specified (or index is set to 0), \
+\n                  the XDP program is attached to all network interfaces.\
+\n                  This parameter is ignored for other program types.\
```
Comment (dthaler) [inline tools/netsh/ebpfnetsh.rc:76]: I think it should instead result in failure if specified for other program types.

File: tools/netsh/ebpfnetsh.rc [hunk 2]
```diff
@@ -111,6 +116,11 @@ BEGIN
 \n                  to detach it.\
 \n      pinpath   - The name to pin the program to, or if not\
 \n                  empty, unpin the program.\
+\n      interface - Optionally the friendly name or index of a \
+\n                  network interface to which an XDP program should be attached. \
+\n                  If the interface is not specified (or index is set to 0), \
+\n                  the XDP program is attached to all network interfaces.\
+\n                  This parameter is ignored for other program types.\
```
Comment (dthaler) [inline tools/netsh/ebpfnetsh.rc:123]: I think it should instead result in failure if specified for other program types.

File: include/ebpf_result.hpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft Corporation
```
Comment (Alan-Jowett) [inline include/ebpf_result.hpp:1]: Why the hpp extension? This is usually reserved for headers with C++ specific features (i.e. that can't be use from a .c file).

File: libs/ebpfnetsh/util.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            row.InterfaceIndex = local_if_index;
+            // Try IPv4 first.
+            row.Family = AF_INET;
+            error = GetIpInterfaceEntry(&row);
+            if (error != ERROR_SUCCESS) {
+                // Try again with IPv6.
+                row.Family = AF_INET;
+                error = GetIpInterfaceEntry(&row);
+            }
+        }
```
Comment (dthaler) [inline libs/ebpfnetsh/util.cpp:48]: You don't need this complexity (and indeed line 45 is a bug).  Just return the ifindex if line 35 is passed.   It may fail later, but no need to check it here too.

File: libs/ebpfnetsh/util.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    } else {
+        // Check if the input string is an interface alias.
+        error = ConvertInterfaceAliasToLuid(arg, &if_luid);
```
Comment (dthaler) [inline libs/ebpfnetsh/util.cpp:51]: for consistency with the "netsh interface ipv4/ipv6" contexts, this should also try [ConvertInterfaceNameToLuidW](https://docs.microsoft.com/en-us/windows/desktop/api/netioapi/nf-netioapi-convertinterfacenametoluidw) as well, not just ConvertInterfaceAliasToLuid.

File: libs/ebpfnetsh/util.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+#pragma once
+
+#include <iostream>
+#include "ebpf_api.h"
+
+std::string
+down_cast_from_wstring(const std::wstring& wide_string);
+
+/**
+ * @brief Parse input string (depicting either interface alias or index) and convert to interface index.
```
Comment (dthaler) [inline libs/ebpfnetsh/util.h:13]:
```suggestion
 * @brief Parse input string (depicting either interface name, alias, or index) and convert to interface index.
```

File: tests/end_to_end/netsh_test.cpp [hunk 7]
```diff
@@ -534,3 +534,53 @@ TEST_CASE("unpin program", "[netsh][programs]")
...
+        handle_ebpf_add_program, &result, 4, L"droppacket.o", L"xdp", L"mypinpath", L"Loopback Pseudo-Interface 1");
+    REQUIRE(strcmp(output.c_str(), "Loaded with ID 196609\n") == 0);
+    REQUIRE(result == NO_ERROR);
+    output = _run_netsh_command(handle_ebpf_delete_program, L"196609", nullptr, nullptr, &result);
+    REQUIRE(result == NO_ERROR);
+    REQUIRE(output == "Unpinned 196609 from mypinpath\n");
+    REQUIRE(bpf_object__next(nullptr) == nullptr);
+
+    // Load program loopback interface index.
+    output = _run_netsh_command(handle_ebpf_add_program, L"droppacket.o", L"xdp", L"interface=1", &result);
```
Comment (dthaler) [inline tests/end_to_end/netsh_test.cpp:556]: besides testing with interface=1, also add tests for "interface=loopback_0" and "interface=Loopback Pseudo-Interface 1".  All three work with "netsh int ip sh <string>" and should similarly work here.

File: tests/end_to_end/netsh_test_helper.h [hunk 8]
```diff
@@ -41,3 +41,6 @@ _run_netsh_command(
     _In_opt_z_ const wchar_t* arg2,
     _In_opt_z_ const wchar_t* arg3,
     _Out_ int* result);
+
+std::string
+run_netsh_command_va_arg(_In_ FN_HANDLE_CMD* command, _Out_ int* result, int argc, ...);
```
Comment (dthaler) [inline tests/end_to_end/netsh_test_helper.h:46]: "va_arg" is probably not quite right since the meaning is slightly different from the va_arg standard api which accepts a va_list as an input parameter. 
```suggestion
run_netsh_command_with_args(_In_ FN_HANDLE_CMD* command, _Out_ int* result, int argc, ...);
```

File: tests/end_to_end/test_helper.cpp [hunk 9]
```diff
@@ -315,12 +315,15 @@ _test_helper_end_to_end::_test_helper_end_to_end()
...
-    REQUIRE(errno == ENOENT);
+    try {
+        // Run down duplicate handles, if any.
+        _duplicate_handles.rundown();
+        // Verify that all maps were successfully removed.
+        uint32_t id;
+        REQUIRE(bpf_map_get_next_id(0, &id) < 0);
+        REQUIRE(errno == ENOENT);
+    } catch (...) {
+    }
```
Comment (dthaler) [inline tests/end_to_end/test_helper.cpp:326]: catching all exceptions (...) is dangerous since it can leave things in an indeterminate state if say a stack exception is caught.

File: tests/end_to_end/netsh_test.cpp (hunk 7 above)
Comment (shankarseal) [reply tests/end_to_end/netsh_test.cpp:556]: `"Loopback Pseudo-Interface 1"` tests is there already.

File: libs/ebpfnetsh/util.cpp (hunk 4 above)
Comment (shankarseal) [reply libs/ebpfnetsh/util.cpp:48]: whoops ! will fix the bug. What about the corner case if a host is IPv6 only?

Comment (dthaler) [reply libs/ebpfnetsh/util.cpp:48]: What about it?  The ifindex is the same.   You don't need to check if it's ipv6 only.  (And indeed there's multiple meanings of the term "ipv6 only")

Comment (dthaler) [reply libs/ebpfnetsh/util.cpp:48]: Lines 37-47 should all be deleted in my view.

Comment (shankarseal) [reply libs/ebpfnetsh/util.cpp:48]: Why shouldn't we check if there is actually an interface with the index passed by user? If IPv4 protocol is not bound to any interface, then `GetIpInterfaceEntry` will fail for AF_INET. Just to clarify `GetIpInterfaceEntry` API requires the address family is specified. Without that specified the API always fails even for legitimate interface index.

File: tests/end_to_end/test_helper.cpp (hunk 9 above)
Comment (shankarseal) [reply tests/end_to_end/test_helper.cpp:326]: Fixed.

File: include/ebpf_result.hpp (hunk 3 above)
Comment (shankarseal) [reply include/ebpf_result.hpp]: Renamed.

File: tools/netsh/ebpfnetsh.rc (hunk 1 above)
Comment (shankarseal) [reply tools/netsh/ebpfnetsh.rc]: Fixed.

File: libs/ebpfnetsh/programs.cpp [hunk 10]
```diff
@@ -141,9 +150,22 @@ handle_ebpf_add_program(
...
     }
 
+    struct bpf_prog_info info;
+    uint32_t info_size = sizeof(info);
+    if (bpf_obj_get_info_by_fd(program_fd, &info, &info_size) < 0) {
+        std::cerr << "error " << errno << ": loaded program but could not get ID" << std::endl;
+        return ERROR_SUPPRESS_OUTPUT;
+    }
+
+    if (info.type == BPF_PROG_TYPE_XDP) {
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/ebpfnetsh/programs.cpp]: Done.

File: include/ebpf_util.h [hunk 11]
```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft Corporation
```
Comment (dthaler) [inline include/ebpf_util.h:1]: eBPF-for-windows coding conventions is to avoid abbreviations.  Can we rename this file to ebpf_utilities.h or something similar?

File: libs/ebpfnetsh/programs.cpp [hunk 12]
```diff
@@ -11,10 +10,10 @@
 #include <netsh.h>
 #include "bpf/bpf.h"
 #include "bpf/libbpf.h"
-#include "ebpf_api.h"
 #include "platform.h"
 #include "programs.h"
 #include "tokens.h"
+#include "util.h"
```
Comment (dthaler) [inline libs/ebpfnetsh/programs.cpp:16]:
```suggestion
#include "utilities.h"
```
avoid abbreviations

File: libs/ebpfnetsh/programs.cpp [hunk 13]
```diff
@@ -47,8 +46,37 @@ static TOKEN_VALUE _ebpf_pinned_type_enum[] = {
...
+ebpf_result_t
+_process_interface_parameter(_In_ LPWSTR interface_parameter, bpf_prog_type prog_type, _Out_ uint32_t* if_index)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    if (_prog_type_supports_interface(prog_type)) {
+        result = parse_ifindex(interface_parameter, if_index);
+        if (result != EBPF_SUCCESS)
+            std::cerr << "Interface parameter is invalid." << std::endl;
+    } else {
+        std::cerr << "Interface parameter is not allowed for program types that dont support interfaces." << std::endl;
```
Comment (dthaler) [inline libs/ebpfnetsh/programs.cpp:64]:
```suggestion
        std::cerr << "Interface parameter is not allowed for program types that don't support interfaces." << std::endl;
```

File: libs/ebpfnetsh/util.cpp [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+            // Try IPv4 first.
+            row.Family = AF_INET;
+            error = GetIpInterfaceEntry(&row);
+            if (error != ERROR_SUCCESS) {
+                // Try again with IPv6.
+                row.Family = AF_INET6;
+                error = GetIpInterfaceEntry(&row);
+            }
+        }
+    } else {
```
Comment (dthaler) [inline libs/ebpfnetsh/util.cpp:49]:
```suggestion
    if ((local_if_index <= 0) || (*end_ptr != L'\0')) {
```
For consistency with other netsh contexts, always treat a positive integer as an ifindex.  Trying to do otherwise would cause confusion due to inconsistency.

File: libs/ebpfnetsh/util.h [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+ * @brief Parse input string (depicting either interface name, alias, or index) and convert to interface index.
+ *
+ * @param[in] arg Input string.
+ * @param[out] if_index Interface index.
+ *
+ * @retval EBPF_SUCCESS Operation succeeded.
+ * @retval EBPF_INVALID_ARGUMENT Input string could not be converted to interface index.
+ */
+ebpf_result_t
+parse_ifindex(_In_z_ const wchar_t* arg, _Out_ uint32_t* if_index);
```
Comment (dthaler) [inline libs/ebpfnetsh/util.h:22]:
```suggestion
parse_ifindex(_In_z_ const wchar_t* arg, _Out_ uint32_t* ifindex);
```
Make arg name match function name (ifindex not if_index).  Also update line 16.

File: libs/ebpfnetsh/util.h [hunk 16]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Microsoft Corporation
```
Comment (dthaler) [inline libs/ebpfnetsh/util.h:1]: Rename util.h and util.cpp to avoid abbreviations.  e.g. "utilities.h" and "utilities.cpp"

File: tests/end_to_end/netsh_test.cpp [hunk 17]
```diff
@@ -534,3 +534,73 @@ TEST_CASE("unpin program", "[netsh][programs]")
...
+    output = run_netsh_command_with_args(
+        handle_ebpf_add_program, &result, 4, L"droppacket.o", L"xdp", L"mypinpath", L"loopback_0");
+    REQUIRE(strcmp(output.c_str(), "Loaded with ID 196610\n") == 0);
+    REQUIRE(result == NO_ERROR);
+    output = _run_netsh_command(handle_ebpf_delete_program, L"196610", nullptr, nullptr, &result);
+    REQUIRE(result == NO_ERROR);
+    REQUIRE(output == "Unpinned 196610 from mypinpath\n");
+    REQUIRE(bpf_object__next(nullptr) == nullptr);
+
+    // Load program loopback interface index.
```
Comment (dthaler) [inline tests/end_to_end/netsh_test.cpp:565]:
```suggestion
    // Load program with loopback interface index.
```

File: tests/end_to_end/netsh_test.cpp [hunk 18]
```diff
@@ -534,3 +534,73 @@ TEST_CASE("unpin program", "[netsh][programs]")
...
+    output = _run_netsh_command(handle_ebpf_add_program, L"droppacket.o", L"xdp", L"interface=1", &result);
+    REQUIRE(strcmp(output.c_str(), "Loaded with ID 196611\n") == 0);
+    REQUIRE(result == NO_ERROR);
+    output = _run_netsh_command(handle_ebpf_delete_program, L"196611", nullptr, nullptr, &result);
+    REQUIRE(result == NO_ERROR);
+    REQUIRE(output == "Unpinned 196611 from DropPacket\n");
+    REQUIRE(bpf_object__next(nullptr) == nullptr);
+
+    // (Negative) Load program incorrect interface name.
```
Comment (dthaler) [inline tests/end_to_end/netsh_test.cpp:574]:
```suggestion
    // (Negative) Load program with incorrect interface name.
```

File: tests/end_to_end/netsh_test.cpp [hunk 19]
```diff
@@ -534,3 +534,73 @@ TEST_CASE("unpin program", "[netsh][programs]")
...
+    // (Negative) Load program with program type that does not support interface parameter.
+    output = _run_netsh_command(handle_ebpf_add_program, L"bindmonitor.o", L"bind", L"interface=1", &result);
+    REQUIRE(
+        strcmp(
+            output.c_str(), "Interface parameter is not allowed for program types that dont support interfaces.\n") ==
+        0);
+    REQUIRE(result == ERROR_SUPPRESS_OUTPUT);
+    REQUIRE(bpf_object__next(nullptr) == nullptr);
+
+    // Add program with no interface parameter
```
Comment (dthaler) [inline tests/end_to_end/netsh_test.cpp:589]:
```suggestion
    // Add program with no interface parameter.
```

File: tests/end_to_end/netsh_test.cpp [hunk 20]
```diff
@@ -534,3 +534,73 @@ TEST_CASE("unpin program", "[netsh][programs]")
...
+    output = _run_netsh_command(handle_ebpf_add_program, L"droppacket.o", nullptr, nullptr, &result);
+    REQUIRE(strcmp(output.c_str(), "Loaded with ID 196614\n") == 0);
+    REQUIRE(result == NO_ERROR);
+    // Detach the program.
```
Comment (dthaler) [inline tests/end_to_end/netsh_test.cpp:593]:
```suggestion

    // Detach the program.
```

File: tools/netsh/ebpfnetsh.rc [hunk 21]
```diff
@@ -69,6 +69,12 @@ BEGIN
 \n                  determine the program type and expected attach\
 \n                  type.  If not specified, the default is xdp.\
 \n      pinpath   - Optionally the name to pin the program to.\
+\n      interface - Optionally the friendly name or index of a \
+\n                  network interface to which an XDP program should be attached. \
+\n                  If the interface is not specified (or index is set to 0), \
```
Comment (dthaler) [inline tools/netsh/ebpfnetsh.rc:74]:
```suggestion
\n      interface - Optionally the friendly name or index of a\
\n                  network interface to which an XDP program should be attached.\
\n                  If the interface is not specified (or index is set to 0),\
```

File: tools/netsh/ebpfnetsh.rc [hunk 22]
```diff
@@ -111,6 +117,12 @@ BEGIN
 \n                  to detach it.\
 \n      pinpath   - The name to pin the program to, or if not\
 \n                  empty, unpin the program.\
+\n      interface - Optionally the friendly name or index of a \
+\n                  network interface to which an XDP program should be attached. \
+\n                  If the interface is not specified (or index is set to 0), \
```
Comment (dthaler) [inline tools/netsh/ebpfnetsh.rc:122]:
```suggestion
\n      interface - Optionally the friendly name or index of a\
\n                  network interface to which an XDP program should be attached.\
\n                  If the interface is not specified (or index is set to 0),\
```
nit: don't need trailing spaces.

File: tools/netsh/ebpfnetsh.rc [hunk 23]
```diff
@@ -69,6 +69,12 @@ BEGIN
 \n                  determine the program type and expected attach\
 \n                  type.  If not specified, the default is xdp.\
 \n      pinpath   - Optionally the name to pin the program to.\
+\n      interface - Optionally the friendly name or index of a \
+\n                  network interface to which an XDP program should be attached. \
+\n                  If the interface is not specified (or index is set to 0), \
+\n                  the XDP program is attached to all network interfaces.\
+\n                  The command fails if this parameter is specified for program types
```
Comment (dthaler) [inline tools/netsh/ebpfnetsh.rc:76]:
```suggestion
\n                  The command fails if this parameter is specified for program types\
```

File: tools/netsh/ebpfnetsh.rc [hunk 24]
```diff
@@ -111,6 +117,12 @@ BEGIN
 \n                  to detach it.\
 \n      pinpath   - The name to pin the program to, or if not\
 \n                  empty, unpin the program.\
+\n      interface - Optionally the friendly name or index of a \
+\n                  network interface to which an XDP program should be attached. \
+\n                  If the interface is not specified (or index is set to 0), \
+\n                  the XDP program is attached to all network interfaces.\
+\n                  The command fails if this parameter is specified for program types
```
Comment (dthaler) [inline tools/netsh/ebpfnetsh.rc:124]:
```suggestion
\n                  The command fails if this parameter is specified for program types\
```

File: libs/ebpfnetsh/util.cpp (hunk 14 above)
Comment (shankarseal) [reply libs/ebpfnetsh/util.cpp]: Changed code to allow +ve integers only, although I had interpreted if_index = 0 as wildcard. But absence of the parameter can mean the same.
I still need to call GetIpInterfceEntry though.

File: tests/end_to_end/netsh_test.cpp (hunk 20 above)
Comment (shankarseal) [reply tests/end_to_end/netsh_test.cpp:593]: Won't fix. The 3 steps, load, detach, re-attach together created one test variation. So I dont want to add line break.

File: libs/ebpfnetsh/util.h (hunk 15 above)
Comment (dthaler) [reply libs/ebpfnetsh/util.h]: This comment still applies to utilities.h

File: libs/ebpfnetsh/util.cpp (hunk 14 above)
Comment (dthaler) [reply libs/ebpfnetsh/util.cpp]: I don't follow.  Why do you need to call GetIpInterfaceEntry?  Other contexts don't.

Comment (shankarseal) [reply libs/ebpfnetsh/util.cpp]: The check was to see if the if_index is valid. But I will move that check to net_ebpf_ext. Otherwise all UM surfaces, netsh, bpftool and libbpf need to replicate the above check.

**Review by dthaler — APPROVED** [review]
Looks good now.  Thanks for addressing my feedback!

**Review by Alan-Jowett — APPROVED** [review]
